src/hooks/useMoveThoughtAnimation.ts

```diff
@@ -0,0 +1,150 @@
+import { throttle } from 'lodash'
+import { useLayoutEffect, useMemo, useState } from 'react'
+import { useSelector } from 'react-redux'
+import State from '../@types/State'
+import durations from '../durations.config'
+import attributeEquals from '../selectors/attributeEquals'
+import usePrevious from './usePrevious'
+
+/**
+ * Determines the move animation type for a thought based on the previous and current indices and the last move action.
+ * Returns:
+ * - 'moveThoughtOver' when this thought is the one explicitly moved by the user (appears on top).
+ * - 'moveThoughtUnder' when this thought is displaced by another thought moving past it (appears underneath).
+ * - null when the movement does not require an animation (e.g. indices unchanged or lastMoveAction is null).
+ */
+const getMoveType = (
+  lastMoveAction: 'moveThoughtUp' | 'moveThoughtDown' | null,
+  previousIndex: number | undefined,
+  currentIndex: number,
+): 'moveThoughtOver' | 'moveThoughtUnder' | null => {
+  if (!lastMoveAction || previousIndex === undefined) return null
+
+  // Determine actual movement direction of this node relative to its previous on-screen index.
+  const direction: 'up' | 'down' = currentIndex < previousIndex ? 'up' : 'down'
+
+  const isMovingWithCursor =
+    (lastMoveAction === 'moveThoughtUp' && direction === 'up') ||
+    (lastMoveAction === 'moveThoughtDown' && direction === 'down')
+
+  return isMovingWithCursor ? 'moveThoughtOver' : 'moveThoughtUnder'
+}
+
+/**
+ * Encapsulates the logic that determines the animation state when a thought is moved
+ * up or down via the moveThoughtUp / moveThoughtDown reducers. This hook abstracts all
+ * of the Redux selectors and timing needed so that the consuming component (TreeNode)
+ * only has to deal with the resulting animation flags.
+ */
+const useMoveThoughtAnimation = (
+  index: number,
+):
+  | {
+      transformOrigin?: React.CSSProperties['transformOrigin']
+      animationDuration?: React.CSSProperties['animationDuration']
+      animationTimingFunction?: React.CSSProperties['animationTimingFunction']
+      animationFillMode?: React.CSSProperties['animationFillMode']
+      animationName?: React.CSSProperties['animationName']
+      zIndex?: React.CSSProperties['zIndex']
+    }
+  | undefined => {
+  const lastMoveAction = useSelector((state: State) => {
+    const lastPatches = state.undoPatches[state.undoPatches.length - 1]
+
+    // Determine if the last action was a moveThoughtUp or moveThoughtDown.
+    const moveType = lastPatches?.some(patch => ['moveThoughtUp', 'moveThoughtDown'].includes(patch.actions[0]))
```

Comment:
Consistent with our renaming, this should be called `moveAction`.

Reply:
Variables have been renamed as suggested. Thanks.